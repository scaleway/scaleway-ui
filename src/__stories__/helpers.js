export const showAlert = (text = 'fired!') => {
  // eslint-disable-next-line no-alert
  alert(text)
}

export const log = (text = 'fired!') => {
  // eslint-disable-next-line no-console
  console.log(text)
}
