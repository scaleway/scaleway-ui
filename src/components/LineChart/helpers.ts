import { Theme } from '@emotion/react'
import { DatumValue, Serie } from '@nivo/line'

const parse = (data?: DatumValue | null) => {
  if (typeof data === 'number') return data
  if (typeof data === 'string') return parseFloat(data)
  if (data instanceof Date) return data.getTime()

  return 0
}

export const getMaxChartValue = (preppedData?: Serie[]): number => {
  if (!preppedData?.length) return 0

  const maximum = Math.max(
    ...preppedData.map(data =>
      data.data.reduce(
        (max, { y }) => Math.max(parse(y), max),
        Number.NEGATIVE_INFINITY,
      ),
    ),
  )

  return Math.ceil(maximum + maximum * 0.1)
}

export const getMinChartValue = (preppedData?: Serie[]): number => {
  if (!preppedData?.length) return 0

  const minimum = Math.min(
    ...preppedData.map(data =>
      data.data.reduce(
        (min, { y }) => Math.min(parse(y), min),
        Number.POSITIVE_INFINITY,
      ),
    ),
  )

  return Math.floor(minimum - minimum * 0.1)
}

export const getMin = (values: number[] = []): number =>
  values.length ? Math.min(...values) : 0

export const getMax = (values: number[] = []): number =>
  values.length ? Math.max(...values) : 0

export const getAverage = (values: number[] = []): number =>
  values.length
    ? Math.round(
        (values.reduce((average, curr) => average + curr, 0) / values.length) *
          100,
      ) / 100
    : 0

export const getCurrent = (values: number[] = []): number =>
  values.length ? values[values.length - 1] : 0

export const getSelected = (
  label: string,
  index: number,
  selected: string[] = [],
): string[] => {
  const labelIndexed = label + index.toString()
  const found = selected.indexOf(labelIndexed)

  if (found > -1) {
    selected.splice(found, 1)
  } else {
    selected.push(labelIndexed)
  }

  return selected
}

export const legendColors = (theme: Theme): string[] =>
  [
    theme.colors.chartGreen,
    theme.colors.chartPurple,
    theme.colors.gray350,
  ] as string[]

export const getLegendColor = (index: number, theme: Theme): string => {
  const colors = legendColors(theme)

  return colors[index] || colors[colors.length - 1]
}
