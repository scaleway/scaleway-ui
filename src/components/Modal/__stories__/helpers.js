export const animations = [
  'fold',
  'zoom',
  'scaleUp',
  'scaleBack',
  'sketch',
  'slideTop',
  'slideBottom',
  'slideLeft',
  'slideRight',
]

export const widths = ['xxsmall', 'xsmall', 'small', 'medium', 'large']

export const placements = [
  'center',
  'top-left',
  'top',
  'top-right',
  'right',
  'bottom-right',
  'bottom',
  'bottom-left',
  'left',
]
