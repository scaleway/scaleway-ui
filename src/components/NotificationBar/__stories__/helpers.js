export default [
  'security',
  'info',
  'success',
  'warning',
  'warning-blue',
  'warning-orange',
  'alert-orange',
  'alert-red',
]
