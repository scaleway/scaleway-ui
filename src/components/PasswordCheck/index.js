import { css } from '@emotion/react'
import PropTypes from 'prop-types'
import React from 'react'
import { Col, Grid, Row } from '../Grid'
import Icon from '../Icon'

const styles = {
  condition: css`
    display: flex;
    align-items: center;
    margin: 0 0 8px 0;
    font-size: 14px;
  `,
  icon: css`
    margin-right: 8px;
    align-self: center;
  `,
}

function PasswordCheck({ rules }) {
  return (
    <Grid fluid gutter={0}>
      <Row>
        {rules.map(rule => (
          <Col large={6} key={rule.name}>
            <p css={styles.condition}>
              <Icon
                name={
                  rule.valid ? 'check-circle-outline' : 'close-circle-outline'
                }
                color={rule.valid ? 'success' : 'darkGrey'}
                css={styles.icon}
                size={20}
              />
              {rule.text}
            </p>
          </Col>
        ))}
      </Row>
    </Grid>
  )
}

PasswordCheck.propTypes = {
  rules: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired,
      valid: PropTypes.bool.isRequired,
    }),
  ).isRequired,
}

export default PasswordCheck
