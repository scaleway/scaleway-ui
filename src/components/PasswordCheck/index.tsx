import { css, useTheme } from '@emotion/react'
import PropTypes from 'prop-types'
import React, { FunctionComponent } from 'react'
import { Col, Grid, Row } from '../Grid'
import Icon from '../Icon'

const styles = {
  condition: css`
    display: flex;
    align-items: center;
    margin: 0 0 8px 0;
    font-size: 14px;
  `,
  icon: css`
    margin-right: 8px;
    align-self: center;
  `,
}

type Rule = {
  name: string
  text: string
  valid: boolean
}

type PasswordCheckProps = {
  /**
   * Is an array of object that contains password rules. `name` is the name of the rule, `text` the text associated
   * with the rule and `valid` is a boolean that determine if the rule is respected or not.
   */
  rules: Rule[]
}

const PasswordCheck: FunctionComponent<PasswordCheckProps> = ({ rules }) => {
  const theme = useTheme()

  return (
    <Grid fluid gutter={0}>
      <Row>
        {rules.map(rule => (
          <Col large={6} key={rule.name}>
            <p css={styles.condition}>
              <Icon
                name={
                  rule.valid ? 'check-circle-outline' : 'close-circle-outline'
                }
                color={
                  rule.valid
                    ? theme.colors.success.textWeak
                    : theme.colors.neutral.textWeak
                }
                css={styles.icon}
                size={20}
              />
              {rule.text}
            </p>
          </Col>
        ))}
      </Row>
    </Grid>
  )
}

PasswordCheck.propTypes = {
  rules: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired,
      valid: PropTypes.bool.isRequired,
    }).isRequired,
  ).isRequired,
}

export default PasswordCheck
