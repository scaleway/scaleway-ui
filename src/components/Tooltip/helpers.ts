import { RefObject } from 'react'

export type TooltipPlacement = 'top' | 'right' | 'bottom' | 'left' | 'auto'
export const ARROW_WIDTH = 6 // in px
export const SPACE = 6 // in px
export const TOTAL_USED_SPACE = ARROW_WIDTH + SPACE // in px
export const DEFAULT_POSITIONS = {
  arrowLeft: 50,
  arrowTop: 99,
  arrowTransform: 'translate(-50%, -50)',
  left: 0,
  rotate: 135,
  tooltipInitialPosition: 'translate3d(0, 0, 0))',
  top: 0,
}

export const TOOLTIP_INITIAL_POSITION = {
  auto: `translate3d(0, ${TOTAL_USED_SPACE}px, 0)`,
  bottom: `translate3d(0, -${TOTAL_USED_SPACE}px, 0)`,
  left: `translate3d(${TOTAL_USED_SPACE}px, 0, 0)`,
  right: `translate3d(-${TOTAL_USED_SPACE}px, 0, 0)`,
  top: `translate3d(0, ${TOTAL_USED_SPACE}px, 0)`,
}

type ComputePlacementTypes = {
  childrenStructuredRef: DOMRect
  tooltipStructuredRef: DOMRect
}

/**
 * This function will find the best placement in a window for tooltip based on children position and tooltip size
 */
export const computePlacement = ({
  childrenStructuredRef,
  tooltipStructuredRef,
}: ComputePlacementTypes) => {
  const {
    top: childrenTop,
    left: childrenLeft,
    right: childrenRight,
  } = childrenStructuredRef

  const { width: tooltipWidth, height: tooltipHeight } = tooltipStructuredRef

  if (childrenTop - tooltipHeight - TOTAL_USED_SPACE < 0) {
    return 'bottom'
  }

  if (childrenLeft - tooltipWidth - TOTAL_USED_SPACE < 0) {
    return 'right'
  }

  if (childrenRight + tooltipWidth + TOTAL_USED_SPACE > window.innerWidth) {
    return 'left'
  }

  return 'top'
}

type ComputePositionsTypes = {
  placement: TooltipPlacement
  childrenRef: RefObject<HTMLDivElement>
  tooltipRef: RefObject<HTMLDivElement>
}

/**
 * This function will compute the positions of tooltip and arrow based on children position and tooltip size
 */
export const computePositions = ({
  placement,
  childrenRef,
  tooltipRef,
}: ComputePositionsTypes) => {
  const childrenStructuredRef = (
    childrenRef.current as HTMLDivElement
  ).getBoundingClientRect()
  const tooltipStructuredRef = (
    tooltipRef.current as HTMLDivElement
  ).getBoundingClientRect()

  const placementBasedOnWindowSize =
    placement === 'auto'
      ? computePlacement({
          childrenStructuredRef,
          tooltipStructuredRef,
        })
      : placement

  const {
    top: childrenTop,
    left: childrenLeft,
    right: childrenRight,
    width: childrenWidth,
    height: childrenHeight,
  } = childrenStructuredRef

  const { width: tooltipWidth, height: tooltipHeight } = tooltipStructuredRef

  switch (placementBasedOnWindowSize) {
    case 'bottom':
      return {
        arrowLeft: tooltipWidth / 2,
        arrowTop: -ARROW_WIDTH - 5,
        arrowTransform: '',
        left: childrenLeft + childrenWidth / 2 - tooltipWidth / 2,
        rotate: 180,
        tooltipInitialPosition: TOOLTIP_INITIAL_POSITION.bottom,
        top: childrenTop + childrenHeight + ARROW_WIDTH + SPACE,
      }
    case 'left':
      return {
        arrowLeft: tooltipWidth + ARROW_WIDTH + 5,
        arrowTop: tooltipHeight / 2,
        arrowTransform: 'translate(-50%, -50%)',
        left: childrenLeft - tooltipWidth - ARROW_WIDTH - SPACE * 2,
        rotate: -90,
        tooltipInitialPosition: TOOLTIP_INITIAL_POSITION.left,
        top: childrenTop - tooltipHeight / 2 + childrenHeight / 2,
      }
    case 'right':
      return {
        arrowLeft: -ARROW_WIDTH - 5,
        arrowTop: tooltipHeight / 2,
        arrowTransform: 'translate(50%, -50%)',
        left: childrenRight + ARROW_WIDTH + SPACE * 2,
        rotate: 90,
        tooltipInitialPosition: TOOLTIP_INITIAL_POSITION.right,
        top: childrenTop - tooltipHeight / 2 + childrenHeight / 2,
      }
    default: // top placement is default value
      return {
        arrowLeft: tooltipWidth / 2,
        arrowTop: tooltipHeight - 1,
        arrowTransform: '',
        left: childrenLeft + childrenWidth / 2 - tooltipWidth / 2,
        rotate: 0,
        tooltipInitialPosition: TOOLTIP_INITIAL_POSITION.top,
        top: childrenTop - tooltipHeight - ARROW_WIDTH - SPACE,
      }
  }
}
