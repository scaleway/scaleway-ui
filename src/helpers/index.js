export * from './clipboard'

export const parseIntOr = (str, fallback) => {
  const value = Number.parseInt(str, 10)
  return !Number.isNaN(value) ? value : fallback
}

export const onKeyOnlyNumbers = e => {
  // `e.which` and `e.keyCode` are deprecated (https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent)
  const keyCode = e.key.charCodeAt(0)
  if (keyCode < 48 || keyCode > 57) {
    e.preventDefault()
  }
}
