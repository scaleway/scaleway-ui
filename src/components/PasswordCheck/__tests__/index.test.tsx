import PasswordCheck from '..'
import { shouldMatchEmotionSnapshot } from '../../../../.jest/helpers'

describe('PasswordCheck', () => {
  test('render with custom values', () =>
    shouldMatchEmotionSnapshot(
      <PasswordCheck
        rules={[
          {
            name: 'custom1',
            text: "That's a beautiful custom password check we have right there",
            valid: false,
          },
          {
            name: 'custom2',
            text: "That's a second beautiful custom password check we have right there",
            valid: true,
          },
        ]}
      />,
    ))
})
