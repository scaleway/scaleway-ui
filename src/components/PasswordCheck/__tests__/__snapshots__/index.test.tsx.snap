// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordCheck render with custom values 1`] = `
<DocumentFragment>
  .cache-1pc1kgu {
  width: 100%;
  margin-left: auto;
  margin-right: auto;
  padding-left: 0;
  padding-right: 0;
}

.cache-1561i4f {
  box-sizing: border-box;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: -8px;
  margin-right: -8px;
}

.cache-1w6xz3u {
  box-sizing: border-box;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  max-width: 100%;
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-left: 8px;
  padding-right: 8px;
}

@media (min-width: 992px) {
  .cache-1w6xz3u {
    -webkit-flex: 0 0 50%;
    -ms-flex: 0 0 50%;
    flex: 0 0 50%;
    max-width: 50%;
  }
}

.cache-13m6wm9-PasswordCheck {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0 0 8px 0;
  font-size: 14px;
}

.cache-c8tbch-PasswordCheck {
  fill: #b2b6c3;
  height: 20px;
  width: 20px;
  min-width: 20px;
  min-height: 20px;
  margin-right: 8px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.cache-c8tbch-PasswordCheck.cache-c8tbch-PasswordCheck {
  vertical-align: middle;
}

.cache-115fstu-PasswordCheck {
  fill: #45d6b5;
  height: 20px;
  width: 20px;
  min-width: 20px;
  min-height: 20px;
  margin-right: 8px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.cache-115fstu-PasswordCheck.cache-115fstu-PasswordCheck {
  vertical-align: middle;
}

<div
    class="em1vjd40 cache-1pc1kgu e7ah3bn0"
  >
    <div
      class="e1uuye5o0 cache-1561i4f e7ah3bn0"
    >
      <div
        class="efyj23m0 cache-1w6xz3u e7ah3bn0"
      >
        <p
          class="cache-13m6wm9-PasswordCheck"
        >
          <svg
            class="etwatq50 cache-c8tbch-PasswordCheck e7ah3bn0"
            viewBox="0 0 24 24"
          >
            <path
              d="M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2C6.47,2 2,6.47 2,12C2,17.53 6.47,22 12,22C17.53,22 22,17.53 22,12C22,6.47 17.53,2 12,2M14.59,8L12,10.59L9.41,8L8,9.41L10.59,12L8,14.59L9.41,16L12,13.41L14.59,16L16,14.59L13.41,12L16,9.41L14.59,8Z"
            />
          </svg>
          That's a beautiful custom password check we have right there
        </p>
      </div>
      <div
        class="efyj23m0 cache-1w6xz3u e7ah3bn0"
      >
        <p
          class="cache-13m6wm9-PasswordCheck"
        >
          <svg
            class="etwatq50 cache-115fstu-PasswordCheck e7ah3bn0"
            viewBox="0 0 24 24"
          >
            <path
              d="M12,2A10,10 0 0,1 22,12A10,10 0 0,1 12,22A10,10 0 0,1 2,12A10,10 0 0,1 12,2M12,4A8,8 0 0,0 4,12A8,8 0 0,0 12,20A8,8 0 0,0 20,12A8,8 0 0,0 12,4M11,16.5L6.5,12L7.91,10.59L11,13.67L16.59,8.09L18,9.5L11,16.5Z"
            />
          </svg>
          That's a second beautiful custom password check we have right there
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
